api/auth/aws/aws.go

```diff
@@ -182,7 +182,7 @@ func (a *AWSAuth) Login(ctx context.Context, client *api.Client) (*api.Secret, e
 	}
 
 	path := fmt.Sprintf("auth/%s/login", a.mountPath)
-	resp, err := client.Logical().Write(path, loginData)
+	resp, err := client.Logical().WriteWithContext(ctx, path, loginData)
```

Comment:
Hi there, I stumbled in this PR when I was looking at https://github.com/hashicorp/vault/pull/14725 
Figured to add a comment here. It seems that the context in these functions is not used before this change. I think it is better to check if the ctx is nil first, and if so, assign a background context to it before passing it to the `WriteWithContext`.  This prevents a possible panic downstream in the context package.
```
if ctx == nil {
		ctx = context.Background()
}
```

Reply:
That's a good point. However, I don't think it's very common (or correct) pass `nil` to any function that expects context. Would it be more idiomatic to return an error in this case?

```go
if ctx == nil {
	return nil, fmt.Errorf("nil context")
}
```

This is similar to what the go standard library is doing in [net/http RequestWithContext](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/net/http/request.go;l=866-868)

Reply:
Well, I am not sure this practice is uncommon, especially when dealing with api code. There are numerous examples of that pattern. Example:
https://github.com/googleapis/google-api-go-client/blob/9d43e7dc08e2598d4dcb661324c4877dfd63c4f9/gmail/v1/gmail-gen.go#L3835

If we were dealing with the server code, I would agree that a nil context should be treated with returning an error. However, this is an api code, and unfortunately, the context has not been used before in there. 

With regards to the net/http package, what I understand is that the function requires a non-nil context by design and it is not necessarily a pattern to follow everywhere. What do you think?

Reply:
Looking at the gmail example, I see that the context is not passed directly to the method, rather it is initialized in a different method altogether [here](https://github.com/googleapis/google-api-go-client/blob/9d43e7dc08e2598d4dcb661324c4877dfd63c4f9/gmail/v1/gmail-gen.go#L3749-L3752). In that code it seems reasonable to check if the context was initialized for the object since the caller may have forgotten to set it.

In the `Login` methods, `ctx context.Context` is an explicit function parameter (effectively part of the function contract). To me, it seems like the `Login` functions should not work with a `nil` context and should require one (either return an error or panic). However, I'm not familiar with the pattern of nil-checking the context that was passed to the function so I'm trying to find the most idiomatic and common solution, particularly for API wrappers. Do you have other examples that we can look at?

Reply:
After some thinking and internal discussions, I'm going to go with your original suggestion just to keep the function fully backwards compatible. We can consider removing the extra nil check in future iterations of this library.

Reply:
Actually, you raised very good point. I am just interested in learning about idiomatic ways of writing code as well, as I am fairly new to GO. I found this example that is kind of interesting. The ctx is passed in the func, and a nil check is done to decide what to do with the request:
https://github.com/googleapis/google-api-go-client/blob/9d43e7dc08e2598d4dcb661324c4877dfd63c4f9/internal/gensupport/send.go#L61

https://github.com/googleapis/google-api-go-client/blob/9d43e7dc08e2598d4dcb661324c4877dfd63c4f9/internal/gensupport/send.go#L27

Reply:
From what I've seen, the most common way of handling contexts passed directly to a function is to just assume that it is never `nil`. Any `nil` contexts would cause a panic once the context is actually used. Here is another example from standard library where context is not nil checked: [database/sql](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/database/sql/sql.go;l=1842).

However, since the Login functions never actually used the context, it seems OK to add a nil check for backwards compatibility. If these were new functions, I would probably lean towards not checking the context for nil at all.